File: cpp/n235_CommonAncestor.py
# https://leetcode-cn.com/problems/lowest-common-ancestor-of-a-binary-search-tree/

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        root_val = root.val
        l_val = min(p.val, q.val)
        r_val = max(p.val, q.val)
        if l_val <= root_val and r_val >= root_val:
            return root
        if l_val < root_val and r_val < root_val:
            return self.lowestCommonAncestor(root.left, p, q)
        if l_val > root_val and r_val > root_val:
            return self.lowestCommonAncestor(root.right, p, q)
            

if __name__ == "__main__":
    pass             

File: cpp/n235_CommonAncestor.cpp
// https://leetcode-cn.com/problems/lowest-common-ancestor-of-a-binary-search-tree/
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;
// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        int root_val = root -> val;
        int left_val = min(p -> val, q -> val);
        int right_val = max(p -> val, q -> val);
        if (left_val <= root_val && root_val <= right_val) return root;
        if (left_val < root_val && right_val < root_val) return lowestCommonAncestor(root -> left, p, q);
        if (left_val > root_val && right_val > root_val) return lowestCommonAncestor(root -> right, p, q);
        return NULL;
    }
};

int main(){
    return 0;
}